### ISSUE TITLE
Performance regressions maybe because of big-endian fixes

### ISSUE THREAD

**Issue by solardiz** [magnumripper/JohnTheRipper#3289]
(This issue title is tentative, subject to change if we find out the cause is different.)

It looks like we got major performance regressions for many formats at least on XOP builds somewhere between 2017/09/19 and now. raw-md4, raw-md5, raw-sha256, raw-sha512 all became several times slower (e.g., 40M to <14M for raw-md5 on one core in FX-8120). md5crypt became somewhat slower (230k+ to <200k). However, sha256crypt and sha512crypt are somehow unaffected.

I suspect the changes for #2861 #2868 could have caused this.

Reviewing the changes that are part of commits 3f450bd2695f578384791a79592bd515b189c740 and 85dfa3c801028bf0e960bf7acab8d25be0ea5fd7 affecting simd-intrinsics.c I see some things that look bad to me and others that look weird to me.

The bad ones are enabling more pieces of code to compile in on little-endian that we had before. Some `#if !ARCH_LITTLE_ENDIAN` and `#endif` lines are removed, and instead #if/#else/#endif blocks are inserted inside of those blocks. This probably creates unneeded overhead on little-endian.

The weird ones are `#if __AVX512F__ || __MIC__` blocks and code specific to those archs inside of upper-level `#if !ARCH_LITTLE_ENDIAN`. Such code should never get compiled in because those archs are little-endian.

While I greatly appreciate Jim's work on the fixes for big-endian, it looks like those were implemented without checking for possible performance regressions on little-endian. We need to revisit that ASAP, as the performance regressions look really bad to me.

Also, some of the performance regressions are so bad that I suspect we could do a lot better than we currently do on big-endian as well - e.g., MD5 doesn't have to be 3 times slower on big-endian than it was on little-endian. I suspect that similar overhead to what we see in these regressions is currently incurred in our big-endian builds. And if so, that overhead looks excessive to me.

Edit: I now also suspect the MAX_KEYS_PER_CRYPT and OMP_SCALE changes and the dropping of `-march=native`. We need to look into all of these potential causes.

**Comment by solardiz** [comment]
With `--enable-native-march`, md5crypt seems to have improved from <200k before to 214k now - but it was 232k+ with last year's code. The fast hashes are still almost as bad as before (even if maybe 10% less bad) - still several times slower than last year. So we'll need to look into what we can do about avoiding the regression from lack of `-march=native` after we've dealt with the bigger fast hash regression maybe from the BE fixes.

**Comment by claudioandre** [comment]
@magnumripper, I added a new label "Regression", and I'll add it to this and to #3122 issues.

**Comment by solardiz** [comment]
It's weird that raw-sha256 became almost 4 times slower (11.8M to 3.0M on one core), but sha256crypt is unaffected. And it's similar for raw-sha512 vs. sha512crypt. The OMP_SCALE changes might also be causing these regressions.

**Comment by solardiz** [comment]
Testing some builds I happen to have made on super:

```
$ for v in bleeding-jumbo-201[78]????; do echo $v; $v/run/john -test -form=raw-md5; done 2> /dev/null
bleeding-jumbo-20170122
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	39098K c/s real, 39098K c/s virtual

bleeding-jumbo-20170515
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	35182K c/s real, 35182K c/s virtual

bleeding-jumbo-20170814
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	39785K c/s real, 39785K c/s virtual

bleeding-jumbo-20171108
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	37824K c/s real, 37824K c/s virtual

bleeding-jumbo-20180119
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	37668K c/s real, 37668K c/s virtual

bleeding-jumbo-20180307
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	20770K c/s real, 20770K c/s virtual

bleeding-jumbo-20180327
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	19199K c/s real, 19199K c/s virtual

bleeding-jumbo-20180422
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	20629K c/s real, 20629K c/s virtual

bleeding-jumbo-20180520
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	19132K c/s real, 19132K c/s virtual

bleeding-jumbo-20180618
Benchmarking: Raw-MD5 [MD5 128/128 AVX 4x3]... DONE
Raw:	20604K c/s real, 20400K c/s virtual
```

Looks like the regression here is "only" 2x, and most of it happened later the Jim's BE work. We need to find out what happened between January and March.

However, there is also a smaller regression (39.8M to 37.8M) consistent with the timing of Jim's BE work. Since there does appear to be added overhead in the source code when building for LE, this is no surprise and we need to fix it.

**Comment by solardiz** [comment]
Tried changing the multiplier on this line:

```
#define MAX_KEYS_PER_CRYPT              (NBKEYS * 16)
```

in `rawMD5_fmt_plug.c` to different values from 1 to 1024 - there's not that much change, in all cases I got speeds between 19M and 21M on one core. So the 2x regression is probably elsewhere.

Maybe the benchmark started limiting the candidate password count to a lower value?

**Comment by magnumripper** [comment]
Skimming through commits between january and march, one suspect is e5b031ed6. Did you only run benchmarks though? How about comparing a real run between bleeding-jumbo-20180119 and bleeding-jumbo-20180307, that would give us clues.

**Comment by magnumripper** [comment]
Another one could be a5ddcde2 "Fix benchmark speed reporting of very fast formats (eg. LM-opencl w/mask).  This bug was introduced at core merge 844df7388 because the replaced int64 "add32to64()" function assumed unsigned whereas the new code did not."

**Comment by solardiz** [comment]
Thanks, magnum. None of these 3 commits you mentioned look like they'd cause what we're seeing. Yes, I only ran benchmarks, and yes we should compare real runs as well.

**Comment by magnumripper** [comment]
I just tried building January/March binaries (using `git checkout bleeding-jumbo@{2018-01-19}` and `git checkout bleeding-jumbo@{2018-03-07}`) on my MBPr and while I see a benchmark regression (could easily be due to e5b031e, I'll check that soon) but not anywhere near the one you saw.
```
$ ../run/john-2018-01-19 -test -form:raw-md5
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	34408K c/s real, 34408K c/s virtual

$ ../run/john-2018-01-19 -test -form:raw-md5
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	34071K c/s real, 34071K c/s virtual

$ ../run/john-2018-01-19 -test -form:raw-md5
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	33804K c/s real, 33804K c/s virtual
```
```
$ ../run/john-2018-03-07 -test -form:raw-md5
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	33437K c/s real, 33437K c/s virtual

$ ../run/john-2018-03-07 -test -form:raw-md5
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	32172K c/s real, 32497K c/s virtual

$ ../run/john-2018-03-07 -test -form:raw-md5
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	33301K c/s real, 33301K c/s virtual
```

**Comment by magnumripper** [comment]
Real runs
```
$ echo aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa > test.in

$ ../run/john-2018-01-19 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Warning: no OpenMP support for this hash type, consider --fork=8
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 22:44) 0g/s 58479Kp/s 58479Kc/s 58479KC/s 99477777..77777777
Session completed

$ ../run/john-2018-01-19 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Warning: no OpenMP support for this hash type, consider --fork=8
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 22:44) 0g/s 55248Kp/s 55248Kc/s 55248KC/s 99477777..77777777
Session completed
```
```
$ ../run/john-2018-03-07 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Warning: no OpenMP support for this hash type, consider --fork=8
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 22:42) 0g/s 58139Kp/s 58139Kc/s 58139KC/s 99477777..77777777
Session completed

$ ../run/john-2018-03-07 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Warning: no OpenMP support for this hash type, consider --fork=8
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 22:45) 0g/s 56818Kp/s 56818Kc/s 56818KC/s 99477777..77777777
Session completed
```
- Actual speed is significantly better than benchmark here - that's odd. Benchmark should have less overhead than a mask mode run.
- Random fluctuations between runs seem to be larger than the regression I thought I saw.

**Comment by solardiz** [comment]
Maybe we incur unaligned accesses somewhere in some builds/invocations? In my benchmarks, maybe only for newer versions. On your AVX2 builds, maybe only in benchmarks but not in real runs?

**Comment by magnumripper** [comment]
Older version
```
$ ../run/john-2017-01-22 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 23:00) 0g/s 59880Kp/s 59880Kc/s 59880KC/s 19677777..77777777
Session completed

$ ../run/john-2017-01-22 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 23:00) 0g/s 59880Kp/s 59880Kc/s 59880KC/s 19677777..77777777
Session completed

$ ../run/john-2017-01-22 test.in -form:raw-md5 -mask:?d?d?d?d?d?d?d?d
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD5 [MD5 256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 DONE (2018-06-27 23:01) 0g/s 59171Kp/s 59171Kc/s 59171KC/s 19677777..77777777
Session completed
```
OK, the above is a little faster than Jan 2018.

But this:
```
$ ../run/john-2017-01-22 -test -form:raw-md5 
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	75299K c/s real, 75299K c/s virtual

$ ../run/john-2017-01-22 -test -form:raw-md5 
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	74962K c/s real, 74220K c/s virtual

$ ../run/john-2017-01-22 -test -form:raw-md5 
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	74282K c/s real, 74282K c/s virtual
```
I recall we had some very inflated benchmarks for a while there (I recall I almost got 100M for NT on well).

In general, comparisons over these time spans are very hard because lots of things changed.

**Comment by solardiz** [comment]
Those 75M figures are actually realistic for AVX2. (And 100M for NT, too.) This is also consistent with 60M you see in real runs, considering that it includes candidate generation "overhead".

**Comment by magnumripper** [comment]
Current bleeding-jumbo
```
$ for i in 1 2 3 ; do ../run/john-2018-06-27 -test -form:raw-md5 ; done
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	32840K c/s real, 32840K c/s virtual

Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	33719K c/s real, 33385K c/s virtual

Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	33327K c/s real, 32997K c/s virtual
```

Same but with simd-intrinsics.[hc] and pseudo-intrinsics.h from Jan 2017 (f6f252bf)
```
$ for i in 1 2 3 ; do ../run/john-2018-06-27-older-simd -test -form:raw-md5 ; done
Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	34369K c/s real, 34369K c/s virtual

Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	34571K c/s real, 34917K c/s virtual

Benchmarking: Raw-MD5 [MD5 256/256 AVX2 8x3]... DONE
Raw:	34288K c/s real, 34631K c/s virtual
```

**Comment by magnumripper** [comment]
> Reviewing the changes that are part of commits 3f450bd and 85dfa3c affecting simd-intrinsics.c I see some things that look bad to me and others that look weird to me.
> 
> The bad ones are enabling more pieces of code to compile in on little-endian that we had before. Some `#if !ARCH_LITTLE_ENDIAN` and `#endif` lines are removed, and instead `#if/#else/#endif` blocks are inserted inside of those blocks. This probably creates unneeded overhead on little-endian.

I can't see much bad stuff (using eg. `git diff -w f6f252bfe simd-intrinsics.c`). Yes he did add another branch for the case of `SSEi_FLAT_OUT` but that is good functionality. Perhaps we can rework it somehow (a switch?) if it happens to blow some branch prediction or instruction cache or whatever magic.

> The weird ones are `#if __AVX512F__ || __MIC__` blocks and code specific to those archs inside of upper-level #if !ARCH_LITTLE_ENDIAN. Such code should never get compiled in because those archs are little-endian.

I can't spot any such thing, neither by looking at a diff as mentioned nor by looking at current code. I do see the opposite: There are `#if __AVX512F__ || __MIC__` blocks, and in the `#else` block of them there are little/big endian alternatives.

**Comment by magnumripper** [comment]
Did some bisecting - e5b031ed definitely caused a huge benchmark hit for NT CPU format on my macbook - it went from over 90K to less than 30K. That doesn't really make sense, I need to investigate why. Not sure how I could miss it at the time, I was probably focused on OpenCL and assumed it was all cold code anyway.

**Comment by magnumripper** [comment]
This mitigates half of the regression:
```diff
diff --git a/src/bench.c b/src/bench.c
index a4a1a7c2f..4716e1b5c 100644
--- a/src/bench.c
+++ b/src/bench.c
@@ -52,6 +52,7 @@
 #include "config.h"
 #include "common-gpu.h"
 #include "mask.h"
+#include "aligned.h"
 
 #ifndef BENCH_BUILD
 #include "options.h"
@@ -223,6 +224,18 @@ static void bench_install_handler(void)
 #endif
 }
 
+static char *strnzcpy_mangle(char *dst, const char *src, int size, int mangle)
+{
+	char *dptr = dst;
+
+	if (size)
+		while (--size && *src)
+			*dptr++ = *src++ ^ mangle;
+	*dptr = 0;
+
+	return dst;
+}
+
 static void bench_set_keys(struct fmt_main *format,
 	struct fmt_tests *current, int cond)
 {
@@ -240,7 +253,6 @@ static void bench_set_keys(struct fmt_main *format,
 	length = format->params.benchmark_length;
 	for (index = 0; index < format->params.max_keys_per_crypt; index++) {
 		char random_pass[PLAINTEXT_BUFFER_SIZE];
-		int i;
 
 		do {
 			if (!current->ciphertext)
@@ -257,9 +269,8 @@ static void bench_set_keys(struct fmt_main *format,
 				break;
 		} while (1);
 
-		strnzcpy(random_pass, plaintext, sizeof(random_pass));
-		for (i = 0; i < strlen(plaintext); i++)
-			random_pass[i] ^= (index & 127);
+		strnzcpy_mangle(random_pass, plaintext,
+		                sizeof(random_pass), index & 127);
 #ifndef BENCH_BUILD
 		if (options.flags & FLG_MASK_CHK) {
 			random_pass[len] = 0;
```
To make it better we'd need to move it out of the hot loop.

**Comment by magnumripper** [comment]
I committed the above now.

**Comment by solardiz** [comment]
> OK, I see now there were such problems in 3f450bd but they were fixed in 85dfa3c.

Oh, you seem to be right. I was looking at these two commits individually and with little context included, so I didn't notice the second one seems to have fixed the first one's placement of some AVX-512/MIC code in big-endian blocks.

And yes, we need to move the mangling of plaintexts out of the hot loop, or alternatively maybe replace it with an equivalent of what happens in mask mode to have the same performance?

Overall, it's bad (a core john limitation) that our candidate password generation is always scalar even when hashing can be rather wide SIMD now. I guess this is the primary thing that enables non-legacy hashcat to run much faster than we do against fast hashes on CPU - it has candidate password generation in OpenCL anyway, so it uses SIMD on CPU for that. With the current formats interface, we can only workaround that by implementing in-format mask mode support much like we do in many OpenCL formats. I was actually thinking of doing something like that for LM hashes when I first introduced those formats interface changes. The result may even be faster than hashcat's approach since the candidate generation and hashing would be more tightly coupled and specialized, but it'd need to be implemented and maintained for each format individually (or for each group of formats maybe, as we could invoke some shared code from inside a format).

**Comment by magnumripper** [comment]
I see another problem. Non-OpenCL build:
```
Benchmarking: NT [MD4 256/256 AVX2 8x3]... DONE
Raw:	71119K c/s real, 71119K c/s virtual
```

OpenCL build (but same CPU format):
```
Benchmarking: NT [MD4 256/256 AVX2 8x3]... DONE
Raw:	65365K c/s real, 65365K c/s virtual
```

We need to find out what causes that.

**Comment by solardiz** [comment]
> This mitigates half of the regression:

I'm getting 29M at raw-md5 on one core in FX-8120 now. This is in fact half way improvement from <14M yesterday to 40M that we had before. All of this is in builds that also include OpenCL.

**Comment by magnumripper** [comment]
I no longer see any difference between OpenCL and non-OpenCL builds. I'm ignoring that issue for now.

**Comment by magnumripper** [comment]
Then it's assigned to you. This issue is all about self-tests not reflecting actual use the slightest. If you can fix this, fine. I will not.

**Comment by solardiz** [comment]
Here's what I mean regarding `raw-sha256`:

```
[solar@super j]$ for v in bleeding-jumbo-201[78]????; do echo $v; OMP_NUM_THREADS=1 $v/run/john -test -form=raw-sha256; done 2> /dev/null
bleeding-jumbo-20170122
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	9089K c/s real, 8999K c/s virtual

bleeding-jumbo-20170515
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	8519K c/s real, 8519K c/s virtual

bleeding-jumbo-20170814
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	9117K c/s real, 9117K c/s virtual

bleeding-jumbo-20171026
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	9011K c/s real, 9011K c/s virtual

bleeding-jumbo-20171108
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	9023K c/s real, 9023K c/s virtual

bleeding-jumbo-20180119
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	4488K c/s real, 4488K c/s virtual

bleeding-jumbo-20180307
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	3288K c/s real, 3288K c/s virtual

bleeding-jumbo-20180327
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	2718K c/s real, 2718K c/s virtual

bleeding-jumbo-20180422
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	2992K c/s real, 2992K c/s virtual

bleeding-jumbo-20180520
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	2411K c/s real, 2411K c/s virtual

bleeding-jumbo-20180618
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	2886K c/s real, 2886K c/s virtual

bleeding-jumbo-20180805
Benchmarking: Raw-SHA256 [SHA256 128/128 AVX 4x]... DONE
Raw:	3772K c/s real, 3772K c/s virtual
```

**Comment by solardiz** [comment]
Actual cracking, 20171108:

```
[solar@super run]$ OMP_NUM_THREADS=1 ./john -form=raw-sha256 -mask='pa?l?l?l?l?l?l' pw
./john: /usr/local/cuda-8.0/targets/x86_64-linux/lib/libOpenCL.so.1: no version information available (required by ./john)
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-SHA256 [SHA256 128/128 AVX 4x])
Warning: OpenMP is disabled; a non-OpenMP build may be faster
Press 'q' or Ctrl-C to abort, almost any other key for status
password         (?)
1g 0:00:00:14 DONE (2018-08-19 14:29) 0.06784g/s 9026Kp/s 9026Kc/s 9026KC/s patavord..pafeanrd
```

20180805:

```
[solar@super run]$ OMP_NUM_THREADS=1 ./john -form=raw-sha256 -mask='pa?l?l?l?l?l?l' pw
./john: /usr/local/cuda-8.0/targets/x86_64-linux/lib/libOpenCL.so.1: no version information available (required by ./john)
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-SHA256 [SHA256 128/128 AVX 4x])
Warning: OpenMP is disabled; a non-OpenMP build may be faster
Press 'q' or Ctrl-C to abort, almost any other key for status
password         (?)
1g 0:00:00:35 DONE (2018-08-19 14:28) 0.02808g/s 3736Kp/s 3736Kc/s 3736KC/s palsword..pamsword
```

This is consistent with the `--test` speeds. That's a ~2.4x slowdown, and it's for real.

**Comment by magnumripper** [comment]
Oh. I can try doing a fully automated git bisect on raw-sha256 tomorrow.

**Comment by solardiz** [comment]
Thanks, magnum. I suspect that part of the problem could be the changed max_keys_per_crypt, but probably there are many changes contributing to this overall effect.

**Comment by magnumripper** [comment]
Note to self: I used this script, named `bisect.script.sh` in base directory. It found the culprit in a few minutes without my attention.
```sh
#!/bin/bash -e
#
# Run with `git bisect run ./bisect.script.sh`
#
# This script should exit with return code 0 for "good" and 1 for "bad"
# return code 125 is special for "could not test" (eg. "could not build")
#
COMMIT=`git rev-parse HEAD`
cd src
./configure --disable-opencl --disable-cuda --disable-rexgen || exit 125
make -s clean || exit 125
make -sj8 || exit 125

# Here's the test
echo 5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8 > hash
rm ../run/john.pot
cp -av ../run/john ../run/john.$COMMIT
OMP_NUM_THREADS=1 ../run/john -form:raw-sha256 hash -mask:pa?l?l?l?l?l?l 2>result.txt || exit 125
SPEED=`grep -oP '[0-9]+Kp/s' result.txt | sed 's#Kp/s##'`

echo $COMMIT $SPEED | tee -a speeds.txt

if [ $SPEED -ge 18000 ]; then
	exit 0
else
	exit 1
fi
```
At first I used `git checkout bleeding-jumbo@{2017-11-08}` and saw no regression at all. I got some 10500Kp/s just like in HEAD. Then I realized what I actually got with that checkout was e6214ceab, much newer than 2017-11-08 for some reason. Manually picked 1b62bbede instead. It does 19395Kp/s - now we're talking.

7b95c8727f is the culprit. It indeed lowered max key per crypts for single-thread without adjusting max_keys_per_crypt, just like you said.

Adding the option `-tune=auto` or `-tune=report` to a current build's command restores the speed completely as far as I can tell.

This is issue #3091 needing some love - a result of Dhiru's changes in 7b95c8727f. Not just this format, but many - I have fixed a good bunch already but we do have hundreds of formats...

**Comment by solardiz** [comment]
Thanks, magnum. I confirm that there's a lot less of a performance regression for `raw-sha256` when running more than 1 thread, thus when we let OMP_SCALE take effect. However, there's still a small regression (actual cracking, tests similar to what I ran above): 12.3M to 11.9M at 2 threads, 37.0M to 35.8M at 32 threads (with GOMP_CPU_AFFINITY=0-31). Of course, this format scales poorly and would be faster with `--fork`, but nevertheless we don't want a regression and this isn't limited to just this one format. I think this smaller regression might have to do with changes to `simd-intrinsics.c` as I had originally suspected.

**Comment by magnumripper** [comment]
Here's a test. I thought I did something similar before but it doesn't seem to be documented above.

current `HEAD` (best of 3):
```
Will run 4 OpenMP threads
Benchmarking: md5crypt, crypt(3) $1$ [MD5 256/256 AVX2 8x3]... (4xOMP) DONE
Raw:	325632 c/s real, 87300 c/s virtual
```

current `HEAD` but with `simd-intrinsics.[ch]` from `3f450bd^` [aka 2bc10023] (best of 3):
```
Will run 4 OpenMP threads
Benchmarking: md5crypt, crypt(3) $1$ [MD5 256/256 AVX2 8x3]... (4xOMP) DONE
Raw:	324864 c/s real, 87094 c/s virtual
```

In this case it's not a regression but a very slight improvement. Perhaps try this on bull?

**Comment by solardiz** [comment]
I'd rather have an easy way to revert to old benchmark code (a john.conf setting?) and/or compare on actual cracking runs for now.

**Comment by magnumripper** [comment]
On a related note:
While looking into #2273 I noticed Jim messed up royally in 3f450bd26, putting back code blocks that was otherwise disabled with `#if USE_EXPERIMENTAL` (and buggy even at that).

I reverted that now (in dc61d377e) so I was kinda hoping to gain back some regression but it actually resulted in 2% performance loss for md5crypt. This is probably highly dependent on CPU type, `-O3` vs `-O2`, `-march=native` and other things though.

**Comment by magnumripper** [comment]
I have looked closely at all net changes of our intrinsics between September 2017 and up to some week ago (before I started optimizing AVX-512) - and there were NO net changes unless built for BE - **except** a bug that Jim introduced (an extra if/else with dead code that was supposed to be defined out) and that was reverted in dc61d377. This affected MD4 and MD5 only. Here's how I reviewed it (this is after reverting said bug):
```
git diff -w f09bb77a..d2571910 simd-intrinsics.c pseudo_intrinsics.h
```
Absolutely nothing changed there, whatever regression you see is elsewhere.

OTOH I believe the recent SHA-512 fix in a2ce5fe3 should be a significant boost for XOP as well. The previously used code path (suboptimal when we have rotate instructions) has been there ever since October 2015 but was probably a regression for XOP - my bad.

**Comment by solardiz** [comment]
The performance for md5crypt and phpass on XOP is almost back to where it was a year ago now.

20170824:

```
Benchmarking: md5crypt, crypt(3) $1$ [MD5 128/128 XOP 4x2]... (8xOMP) DONE
Raw:    232448 c/s real, 29092 c/s virtual

Benchmarking: phpass ($P$9) [phpass ($P$ or $H$) 128/128 XOP 4x2]... (8xOMP) DONE
Speed for cost 1 (iteration count) of 2048
Raw:    135857 c/s real, 16961 c/s virtual
```

20181013:

```
Benchmarking: md5crypt, crypt(3) $1$ [MD5 128/128 XOP 4x2]... (8xOMP) DONE
Raw:    231936 c/s real, 28992 c/s virtual

Benchmarking: phpass ($P$9) [phpass ($P$ or $H$) 128/128 XOP 4x2]... (8xOMP) DONE
Speed for cost 1 (iteration count) of 2048
Raw:    135168 c/s real, 16896 c/s virtual
```

However, since you say the recently fixed regression for XOP was there since 2015 perhaps this is a cumulative effect of some other still-present regression introduced during the last year and fixing of this older regression. Perhaps we can actually improve speeds compared to last year's by also fixing the new regression.

The `--test` speeds for raw hashes are still weird - e.g., raw-md5 appears faster (30M on one core) than raw-md4 (25M).

**Comment by solardiz** [comment]
It's weird: applying the equivalent change of a2ce5fe33ee1b61ea3f928c2ddd8a8e6561313ac to last year's code does significantly reduce code size, but provides no obvious speedup on FX-8120.

```
$ size simd-intrinsics*.o
   text	   data	    bss	    dec	    hex	filename
  98966	      0	      0	  98966	  18296	simd-intrinsics-hack.o
 101302	      0	      0	 101302	  18bb6	simd-intrinsics-orig.o
```

(This is based on tree from 20170919, where speeds at md5crypt and phpass are essentially the same as shown in the comment above either way.)

**Comment by solardiz** [comment]
Trying to make sense of this: maybe either regression introduced a code size increase, exceeding L1i cache on AMD FX, and fixing either regression alone is sufficient to make the code fit in the cache again.

**Comment by solardiz** [comment]
It is possible that we fixed whatever regression there was with dc61d377eb5fc4d13cba91c653fad7c92247db1f, whereas a2ce5fe33ee1b61ea3f928c2ddd8a8e6561313ac surprisingly doesn't make much of a difference in speed on XOP. I didn't do any builds in September, and am too lazy to checkout a version from inbetween these two commits to check. Either way, it looks like we're all set in terms of performance at md5crypt & phpass on XOP. There's still work left to do on proper max_keys_per_crypt for fast hashes, where I think we still have regressions compared to last year.

**Comment by solardiz** [comment]
I think we took care of issues relating to the big-endian fixes,  AMD FX and XOP, md5crypt and phpass.

What's left is:

1. We need to re-introduce benchmarks that can be directly compared to previous release's. This can be a non-default option only for use by our team.

2. We need to (re-)tune MKPC and OMP_SCALE, especially for fast hashes, along with doing relbench runs against older versions on different machines.

But I feel these should be tracked as part of our next release preparation (OK to create sub-issues for the above) and not here. So closing this issue.
